### ISSUE TITLE
[BUG] 3.2 Alpha2+ - Build Failure on NonStop (again) for missing struct timeval.

### ISSUE THREAD

**Issue by rsbeckerca** [openssl/openssl#22309]
This is a repeat of a prior problem, different location:

```
                  if (!TEST_true(SSL_get_event_timeout(c_ssl, &tv, &isinf)))
                       ^
"/home/jenkinsbuild/.jenkins/workspace/OpenSSL-3.2_Monitor/test/quic_tserver_test.c", line 326: error(252):
          argument of type "struct timeval *" is incompatible with parameter of type "struct timeval *"
```
c99 requires struct definitions to be fully specified when used. The `struct timeval` needs to be included as with what we previously reported in other quic tests.

**Comment by levitte** [comment]
Actually, this is bigger than just NonStop.

The use of `struct timeval` in ssl.h has the unfortunate consequence, that developers that use OpenSSL must now know to include `<sys/time.h>`, `<sys/socket.h>`, `<winsock2.h>`, or other variants I forget (depending on platform) before including `<openssl/ssl.h>`.

We haven't documented this, so will most probably be a surprise for some.

**Comment by t-j-h** [comment]
All of these related struct timeval things were being solved the wrong way using internal headerfile includes.

We have multiple PRs working about a mistake where ssl.h isn't safe to include by itself. We need a header file the application can use to handle the platform differences here and not have apps etc using internal headers to work around this not being solved cleanly.

See https://github.com/openssl/openssl/pull/22310#issuecomment-1752193245

**Comment by levitte** [comment]
Actually, `<openssl/e_os2.h>` is already included by `<openssl/ssl.h>`, so it could be a header to look at.  I think it's been considered before, but the pushback has been that it would suck in socket headers *everywhere* on platforms that get timeval that way, and that was apparently undesirable (I frankly don't understand why...)

**Comment by t-j-h** [comment]
Huh? e_os2.h does not include what is required - so it isn't helpful.

We have a typedef that we need defined in ssl.h and a platform specific dance that has to be done to get the required header file to handle that - a dance we do in a couple of places already what need struct timeval that really should be in one place only.

We have to think on the context of the application developers using OpenSSL and make sure we solve the problem in a manner than doesn't require all applications using ssl.h to have to go and change their code.

**Comment by levitte** [comment]
I meant that e_os2.h could be modified to include what's necessary.

**Comment by t-j-h** [comment]
And if e_os2.h is modified to include winsock2.h on windows then it will impact all users of e_os2.h not just ssl.h - this is a change where any file using struct timeval should be impacted and have the include dance handling in a single location and e_os2.h is most likely the wrong place for this as it is heavily used elsewhere.

The socket related stuff being split out of e_os2.h was one of the main reasons it exists. You could copy all that stuff from e_os.h back into e_os2.h but that isn't what I was suggesting - I suggest fixing this for those contexts where we need the struct timeval item handled - and removing the various other ways we handle that the code so it is in one place.


There is still the separate issue where we have use of winsock.h which should simple not exist in the code - only winsock2.h should be used - and they are not ever meant to be simultaneously used.

**Comment by levitte** [comment]
> And if e_os2.h is modified to include winsock2.h on windows then it will impact all users of e_os2.h not just ssl.h - this is a change where any file using struct timeval should be impacted and have the include dance handling in a single location and e_os2.h is most likely the wrong place for this as it is heavily used elsewhere.

Yeah that was the pushback I mentioned.  So winsock2.h is too heavy for general inclusion, gotcha.

**Comment by t-j-h** [comment]
We do a lot more "interesting" things than winsock2.h header inclusion in e_os.h and we do a similar dance in other places.
That needs cleaning up. There is little harm in pure winsock2.h inclusion - but for people building libcrypto with NO_SOCK they would most likely want to continue to avoid winsock2.h inclusion. There are also issues about having to set certain things before including header files and having winsock2.h kick in before windows.h and other such things.

Look at the dance in bio_sock.c  (which does use plain winsock.h) and internal/sockets.h also using winsock.h which is an error.

**Comment by levitte** [comment]
Note that I don't suggest making everything in e_os.h public.  Most of it is interesting for our internal usage only, as far as I know.

I agree, though, that we'd need a better architecture abstraction.  That's a major effort, though.

Regarding winsock.h vs winsock2.h, an issue for that specifically should be raised.

**Comment by hlandau** [comment]
Perhaps a new `e_ossock.h` as a standard way to get OS-specific socket includes? `ssl.h` could then include it.

**Comment by t8m** [comment]
Fixed by #22323
